README.md


Language: [English](README.md) | [中文简体](README-ZH.md)

# FlutterSocket

[![build status](https://img.shields.io/travis/flutterchina/dio/vm.svg?style=flat-square)](https://waitwalker.cn/2019/08/15/%E4%BA%A4%E6%B5%81%E7%BE%A4%E7%BB%84/)
[![Pub](https://img.shields.io/pub/v/flutter_socket_plugin.svg?style=flat-square)](https://pub.dartlang.org/packages/flutter_socket_plugin)
[![support](https://img.shields.io/badge/platform-flutter%7Cdart%20vm-ff69b4.svg?style=flat-square)](https://waitwalker.cn/2019/08/15/%E4%BA%A4%E6%B5%81%E7%BE%A4%E7%BB%84/)


FlutterSocket is a practical cross-platform socket plugin. At present, it has realized the basic functions of client side: create, connect, send, receive, disconnect and so on.
### If you need other functions, you can expand on this basis.

## Demo 

![Demo](https://github.com/waitwalker/Resources/blob/master/Flutter/FlutterSocket/socket_demo.gif?raw=true)

### Video
video url: https://www.youtube.com/watch?v=fmtwYHeOvE0
<video src="https://www.youtube.com/watch?v=fmtwYHeOvE0" controls="controls" width="500" height="300">Your browser does not support playing this video！</video>

## Add dependency

```yaml
dependencies:
  flutter_socket_plugin: lastest version  
```

## Simple to use

```dart
import 'package:flutter_socket_plugin/flutter_socket_plugin.dart';
  ///
  /// @Method: initSocket
  /// @Parameter:
  /// @ReturnType:
  /// @Description: init socket
  /// @author: waitwalker
  /// @Date: 2019-08-23
  ///
  initSocket() {
    
    /// init socket
    flutterSocket = FlutterSocket();

    /// listen connect callback
    flutterSocket.connectListener((data){
      print("connect listener data:$data");
    });

    /// listen error callback
    flutterSocket.errorListener((data){
      print("error listener data:$data");
    });

    /// listen receive callback
    flutterSocket.receiveListener((data){
      print("receive listener data:$data");
      if (data != null) {
        receiveMessage = receiveMessage + "\n" + data;
      }
      setState(() {

      });
    });

    /// listen disconnect callback
    flutterSocket.disconnectListener((data){
      print("disconnect listener data:$data");
    });

  }
```


### Steps

Create `socket` :

```dart
await flutterSocket.createSocket("192.168.8.120", 10007, timeout: 20);
```

Connect `connect` :

```dart
flutterSocket.tryConnect();
```

Send message `send` :

```dart
flutterSocket.send(textEditingController.text);
```

Disconnect `disconnect` :

```dart
flutterSocket.tryDisconnect();
```


### Examples

you can find all examples code [here](https://github.com/waitwalker/flutter_socket_plugin/tree/master/example).

## Future

FlutterSocket currently implements the sending and receiving of basic String messages on the client. In the future, it will gradually improve its functions if conditions permit. If you need some functions urgently, you can continue to add functions on the basis of FlutterSocket to make the function of FlutterSocket more powerful and perfect.

## Copyright & License

This project is completely open source and the license is MIT. If you like, welcome star.

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/waitwalker/flutter_socket_plugin/issues

## Flutter Technology and Other Communication Groups

wechat:dbzy_duzhao

Scan QR code👇:

<img src="https://github.com/waitwalker/Resources/blob/master/Flutter/group/flutter_development_0923.JPG?raw=true" width="350" height="500" align=center />



CHANGELOG.md
## 0.0.1

Completion of basic plugin functions

## 0.0.2

Update some documents

## 0.0.3

Update some documents

## 0.0.4

Update group

## 1.0.0

Release version

## 1.0.1

Adapt to Swift 5.0+

## 2.0.0

Adapt to Flutter 2.0 & nullsafety

## 3.0.0

Function optimization

example/lib/client_page.dart
import 'package:flutter/material.dart';
import 'package:flutter_socket_plugin/flutter_socket_plugin.dart';

///
/// @description Socket Client sample code
/// @author waitwalker
/// @time 2021/6/30 17:07
///
class ClientPage extends StatefulWidget {
  @override
  _ClientPageState createState() => _ClientPageState();
}

class _ClientPageState extends State<ClientPage> {

  TextEditingController textEditingController = TextEditingController();

  FlutterSocket? flutterSocket;
  bool connected = false;
  String _host = "192.168.10.63";
  int _port = 10007;
  String receiveMessage = "";

  @override
  void initState() {
    initSocket();
    textEditingController.addListener((){
      print("input text:${textEditingController.text}");
    });
    super.initState();
  }

  ///
  /// @Method: initSocket
  /// @Parameter:
  /// @ReturnType:
  /// @Description: init socket
  /// @author: waitwalker
  /// @Date: 2019-08-23
  ///
  initSocket() {

    /// init socket
    flutterSocket = FlutterSocket();

    /// listen connect callback
    flutterSocket!.connectListener((data){
      print("connect listener data:$data");
    });

    /// listen error callback
    flutterSocket!.errorListener((data){
      print("error listener data:$data");
    });

    /// listen receive callback
    flutterSocket!.receiveListener((data){
      print("receive listener data:$data");
      if (data != null) {
        receiveMessage = receiveMessage + "\n" + data;
      }
      setState(() {

      });
    });

    /// listen disconnect callback
    flutterSocket!.disconnectListener((data){
      print("disconnect listener data:$data");
    });
  }


  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("FlutterSocket client",style: TextStyle(fontSize: 18,color: Colors.white),),
        backgroundColor: Colors.lightBlue,
      ),
      body: Column(
        children: <Widget>[
          Expanded(child: ListView(children: [
            Padding(
              padding: EdgeInsets.only(top: 30,left: 20,right: 20),
              child: Row(
                children: <Widget>[
                  Text("Host:",style: TextStyle(fontSize: 22,color: Colors.lightBlueAccent,fontWeight: FontWeight.w600),),
                  Padding(padding: EdgeInsets.only(left: 10),),
                  Container(
                    decoration: BoxDecoration(
                      border: Border(bottom: BorderSide(color: Colors.redAccent,width: 1.0)),
                    ),
                    child: Text(_host,style: TextStyle(fontSize: 18,color: Colors.black87,fontWeight: FontWeight.w300),),
                  ),
                  Padding(padding: EdgeInsets.only(left: 30),),
                  Text("Port:",style: TextStyle(fontSize: 22,color: Colors.lightBlueAccent,fontWeight: FontWeight.w600),),
                  Padding(padding: EdgeInsets.only(left: 10),),
                  Container(
                    decoration: BoxDecoration(
                      border: Border(bottom: BorderSide(color: Colors.redAccent,width: 1.0)),
                    ),
                    child: Text("$_port",style: TextStyle(fontSize: 18,color: Colors.black87,fontWeight: FontWeight.w300),),
                  ),
                ],
              ),
            ),

            Padding(
              padding: EdgeInsets.only(left: 20,right: 20,top: 30),
              child: Row(
                children: <Widget>[
                  ElevatedButton(
                    style: ButtonStyle(
                      backgroundColor: ButtonStyleButton.allOrNull<Color>(Colors.lightBlueAccent),
                    ),
                    child: Text("Create",style: TextStyle(fontSize: 18,color: Colors.white,fontWeight: FontWeight.w500),),
                    onPressed: () async {
                      await flutterSocket!.createSocket(_host, _port, timeout: 20);
                    },
                  ),
                  Padding(padding: EdgeInsets.only(left: 20)),
                  ElevatedButton(
                    style: ButtonStyle(
                      backgroundColor: ButtonStyleButton.allOrNull<Color>(Colors.lightBlueAccent),
                    ),
                    child: Text("Connect",style: TextStyle(fontSize: 18,color: Colors.white,fontWeight: FontWeight.w500),),
                    onPressed: (){
                      flutterSocket!.tryConnect();
                    },
                  ),
                  Padding(padding: EdgeInsets.only(left: 20)),
                  ElevatedButton(
                    style: ButtonStyle(
                      backgroundColor: ButtonStyleButton.allOrNull<Color>(Colors.lightBlueAccent),
                    ),
                    child: Text("Disconnect",style: TextStyle(fontSize: 18,color: Colors.white,fontWeight: FontWeight.w500),),
                    onPressed: (){
                      flutterSocket!.tryDisconnect();
                    },
                  ),
                ],
              ),
            ),

            Padding(
              padding: EdgeInsets.only(left: 20,right: 20,top: 30,),
              child: Container(
                height: 150,
                decoration: BoxDecoration(
                    border: Border.all(color: Colors.lightBlueAccent,width: 1.0)
                ),
                child: TextField(
                  controller: textEditingController,
                  decoration: InputDecoration(
                      labelText: "Input content",
                      border: InputBorder.none
                  ),
                ),
              ),
            ),

            Padding(
              padding: EdgeInsets.only(left: 20,right: 20,top: 10),
              child: Container(
                alignment: Alignment.centerLeft,
                child: ElevatedButton(
                  style: ButtonStyle(
                    backgroundColor: ButtonStyleButton.allOrNull<Color>(Colors.lightBlueAccent),
                  ),
                  child: Text("Send",style: TextStyle(fontSize: 18,color: Colors.white,fontWeight: FontWeight.w500),),
                  onPressed: (){
                    print("will send message:${textEditingController.text}");
                    flutterSocket!.send(textEditingController.text);
                  },),
              ),
            ),

            Padding(
              padding: EdgeInsets.only(left: 20,right: 20,top: 30),
              child: Container(
                alignment: Alignment.centerLeft,
                child: ElevatedButton(
                  style: ButtonStyle(
                    backgroundColor: ButtonStyleButton.allOrNull<Color>(Colors.lightBlueAccent),
                  ),
                  child: Text("Receive Data",style: TextStyle(fontSize: 18,color: Colors.white,fontWeight: FontWeight.w500),),
                  onPressed: (){},),
              ),
            ),

            Padding(
              padding: EdgeInsets.only(left: 20,right: 20,top: 10),
              child: Container(
                height: 150,
                width: MediaQuery.of(context).size.width - 40,
                decoration: BoxDecoration(
                    border: Border.all(color: Colors.lightBlueAccent,width: 1.0)
                ),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.start,
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Expanded(child: Text(receiveMessage,style: TextStyle(fontSize: 15,color: Colors.redAccent, fontWeight: FontWeight.w500),),)
                  ],
                ),
              ),
            ),

            Padding(
              padding: EdgeInsets.only(left: 20,right: 20,top: 10),
              child: Container(
                alignment: Alignment.centerLeft,
                child: ElevatedButton(
                  style: ButtonStyle(
                    backgroundColor: ButtonStyleButton.allOrNull<Color>(Colors.lightBlueAccent),
                  ),
                  child: Text("Clear",style: TextStyle(fontSize: 18,color: Colors.white,fontWeight: FontWeight.w500),),
                  onPressed: (){
                    setState(() {
                      receiveMessage = "";
                    });
                  },),
              ),
            ),
          ],),),
        ],
      ),
    );
  }

}